Issue with SBOM generation in Harbor v2.11.0 when using external Reverse Proxy (HTTP 404)

Issue:
The new SBOM generation feature in version 2.11.0 uses the external URL, leading to 404 errors due to the reverse proxy configuration.

Suggested Fix:
I recommend that the internal server name be used for SBOM uploads like other internal communications to avoid this issue. This would ensure consistency and prevent misdirected requests that result in 404 errors (and connections failures when calling external services).

Question:
Is there a way to configure Harbor to use the container internal server name instead of the external URL when uploading SBOM results?

Thank for your great work and for looking into this issue. I appreciate any guidance or updates you can provide.

Best regards,
Oliver

---

My configuration:

User Access --> https://myrepo.example.com (HTTPS)
   | 
External Nginx Reverse Proxy --> https://internalserver.example.local (HTTPS)
   |
Harbor by docker compose

In version 2.10.x, everything functions as expected, including upload of vulnerability scans. 
However, in version 2.11.x, the new SBOM generation feature seems to be mistakenly using the external URL (https://myrepo.example.com) for internal communications. This misconfiguration leads to a situation where the external reverse proxy only returns HTTP 404 when accessed from the internal network.

docker compose logs
```
harbor-jobservice  | 2024-06-09T13:20:26Z [ERROR] [/pkg/scan/sbom/sbom.go:108]: error when create accessory from image GET https://myrepo.example.com/v2/: unexpected status code 404 Not Found: <html>
harbor-jobservice  | <head><title>404 Not Found</title></head>
harbor-jobservice  | <body>
harbor-jobservice  | <center><h1>404 Not Found</h1></center>
harbor-jobservice  | <hr><center>nginx</center>
harbor-jobservice  | </body>
harbor-jobservice  | </html>
harbor-jobservice  | 2024-06-09T13:20:26Z [ERROR] [/pkg/scan/job.go:307]: Failed to convert vulnerability data to new schema for report 08c76d25-a6e4-4c5c-bc45-1e01260cc224, error GET https://myrepo.example.com/v2/: unexpected status code 404 Not Found: <html>
harbor-jobservice  | <head><title>404 Not Found</title></head>
harbor-jobservice  | <body>
harbor-jobservice  | <center><h1>404 Not Found</h1></center>
harbor-jobservice  | <hr><center>nginx</center>
harbor-jobservice  | </body>
harbor-jobservice  | </html>
harbor-jobservice  | 2024-06-09T13:20:26Z [ERROR] [/jobservice/runner/redis.go:123]: Job 'IMAGE_SCAN:6c49933c8512c5f47b6739c7' exit with error: run error: GET https://myrepo.example.com/v2/: unexpected status code 404 Not Found: <html>
harbor-jobservice  | <head><title>404 Not Found</title></head>
harbor-jobservice  | <body>
harbor-jobservice  | <center><h1>404 Not Found</h1></center>
harbor-jobservice  | <hr><center>nginx</center>
harbor-jobservice  | </body>
harbor-jobservice  | </html>
```

harbor.yml
```
hostname: myrepo.example.com
http:
  port: 80
https:
  port: 443
  certificate: /root/owa/harbor_install/harbor/data/cert/internalserver.example.local.crt
  private_key: /root/owa/harbor_install/harbor/data/cert/internalserver.example.local.key

# internal_tls:
#   # set enabled to true means internal tls is enabled
#   enabled: true

# external_url: https://myrepo.example.com:443
# test only
# external_url: https://nginx:8443
```

I tried to switch my configuration to ```external_url: https://nginx:8443``` and the sbom upload is working, but docker pull from https://myrepo.example.com failed with ```dial tcp: lookup nginx on ...: server misbehaving```

--- a/src/pkg/scan/sbom/sbom.go
+++ b/src/pkg/scan/sbom/sbom.go
@@ -22,13 +22,11 @@ import (
 	"strings"
 	"time"
 
-	"github.com/goharbor/harbor/src/common"
 	"github.com/goharbor/harbor/src/common/rbac"
 	"github.com/goharbor/harbor/src/controller/artifact"
 	scanCtl "github.com/goharbor/harbor/src/controller/scan"
 	"github.com/goharbor/harbor/src/jobservice/job"
 	"github.com/goharbor/harbor/src/jobservice/logger"
-	"github.com/goharbor/harbor/src/lib/config"
 	"github.com/goharbor/harbor/src/lib/errors"
 	"github.com/goharbor/harbor/src/lib/log"
 	"github.com/goharbor/harbor/src/lib/orm"
@@ -54,7 +52,6 @@ const (
 func init() {
 	scan.RegisterScanHanlder(v1.ScanTypeSbom, &scanHandler{
 		GenAccessoryFunc:       scan.GenAccessoryArt,
-		RegistryServer:         registry,
 		SBOMMgrFunc:            func() Manager { return Mgr },
 		TaskMgrFunc:            func() task.Manager { return task.Mgr },
 		ArtifactControllerFunc: func() artifact.Controller { return artifact.Ctl },
@@ -67,7 +64,6 @@ func init() {
 // scanHandler defines the Handler to generate sbom
 type scanHandler struct {
 	GenAccessoryFunc       func(scanRep v1.ScanRequest, sbomContent []byte, labels map[string]string, mediaType string, robot *model.Robot) (string, error)
-	RegistryServer         func(ctx context.Context) (string, bool)
 	SBOMMgrFunc            func() Manager
 	TaskMgrFunc            func() task.Manager
 	ArtifactControllerFunc func() artifact.Controller
@@ -96,8 +92,10 @@ func (h *scanHandler) PostScan(ctx job.Context, sr *v1.ScanRequest, _ *scanModel
 		Registry: sr.Registry,
 		Artifact: sr.Artifact,
 	}
-	// the registry server url is core by default, need to replace it with real registry server url
-	scanReq.Registry.URL, scanReq.Registry.Insecure = h.RegistryServer(ctx.SystemContext())
+	scanReq.Registry.Insecure = strings.HasPrefix(scanReq.Registry.URL, "http://")
+	// the registry URL should not contain http:// or https:// prefix
+	scanReq.Registry.URL = strings.TrimPrefix(scanReq.Registry.URL, "http://")
+	scanReq.Registry.URL = strings.TrimPrefix(scanReq.Registry.URL, "https://")
 	if len(scanReq.Registry.URL) == 0 {
 		return "", fmt.Errorf("empty registry server")
 	}
@@ -168,19 +166,6 @@ func (h *scanHandler) Update(ctx context.Context, uuid string, report string) er
 		return err
 	}
 	return nil
-}
-
-// extract server name from config, and remove the protocol prefix
-func registry(ctx context.Context) (string, bool) {
-	cfgMgr, ok := config.FromContext(ctx)
-	if ok {
-		extURL := cfgMgr.Get(context.Background(), common.ExtEndpoint).GetString()
-		insecure := strings.HasPrefix(extURL, "http://")
-		server := strings.TrimPrefix(extURL, "https://")
-		server = strings.TrimPrefix(server, "http://")
-		return server, insecure
-	}
-	return "", false
 }
 
 // retrieveSBOMContent retrieves the "sbom" field from the raw report

--- a/src/pkg/scan/sbom/sbom_test.go
+++ b/src/pkg/scan/sbom/sbom_test.go
@@ -147,7 +147,6 @@ func (suite *SBOMTestSuite) SetupSuite() {
 
 	suite.handler = &scanHandler{
 		GenAccessoryFunc:       mockGenAccessory,
-		RegistryServer:         mockGetRegistry,
 		SBOMMgrFunc:            func() Manager { return suite.sbomManager },
 		TaskMgrFunc:            func() task.Manager { return suite.taskMgr },
 		ArtifactControllerFunc: func() artifact.Controller { return suite.artifactCtl },

